server/storage/storage.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/encryptionkm"
+	"github.com/tikv/pd/server/kv"
+	"github.com/tikv/pd/server/storage/endpoint"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// Storage is the interface for the backend storage of the PD.
+type Storage interface {
+	// Introducing the kv.Base here is to provide
+	// the basic key-value read/write ability for the Storage.
+	kv.Base
+	endpoint.ConfigStorage
+	endpoint.MetaStorage
+	endpoint.RuleStorage
+	endpoint.ComponentStorage
+	endpoint.ReplicationStatusStorage
+	endpoint.GCSafePointStorage
+}
+
+// NewStorageWithMemoryBackend creates a new storage with memory backend.
+func NewStorageWithMemoryBackend() Storage {
+	return newMemoryBackend()
+}
+
+// NewStorageWithEtcdBackend creates a new storage with etcd backend.
+func NewStorageWithEtcdBackend(client *clientv3.Client, rootPath string) Storage {
+	return newEtcdBackend(client, rootPath)
+}
+
+// NewStorageWithLevelDBBackend creates a new storage with LevelDB backend.
+func NewStorageWithLevelDBBackend(
+	ctx context.Context,
+	rootPath string,
+	ekm *encryptionkm.KeyManager,
+) (*levelDBBackend, error) {
+	return newLevelDBBackend(ctx, rootPath, ekm)
+}
+
+// TODO: support other KV storage backends like BadgerDB in the future.
+
+type pdStorage struct {
```

Comment:
prefer to use another name

Reply:
This storage contains both etcd and LevelDB to save meta and region info in most cases. I think it's OK to make its name start with "pd" to indicate the special use by the `pd-server`.